versionedRouter.js

```diff
@@ -295,7 +295,13 @@ async function handleDest(ctx, version, destination) {
 
             return {
               output: { ...ev, userId },
-              metadata: event.metadata,
+              metadata: destHandler.processMetadata
+                ? destHandler.processMetadata({
+                    metadata: event.metadata,
+                    inputEvent: parsedEvent,
+                    outputEvent: ev
+                  })
+                : event.metadata,
```

Comment:
Why can't we manipulate the `metadata` within the `process` method itself? We have access to all the data there.

Reply:
No it will break the contact between server and transformer if you check carefully then we have to pass updated `metadata` in `events` array which will break the structure of `output`

Reply:
How does it break the data contract? You'd send `metadata` for each event processed, and the value of `rudderId` is only updated now.

Reply:
Currently we don't sent metadata as part of the processed output from `process` function

Reply:
```
 output: { ...ev, userId }
```
each of the events `ev`s would then contain `metadata`

Reply:
We don't have to return anything from the process function. As you have the same metadata object reference in the function, updating it would be sufficient.

Reply:
That would lead to some level of abstraction and unwanted hiding of context. Then `process` function implicitly has a side effect i.e it does something more (mutates) state of an object which it isn't supposed to

Reply:
I see your point.